File: src/main/java/org/moinex/ui/dialog/investment/ArchivedBondsController.java
/*
 * Filename: ArchivedBondsController.java
 * Created on: January  3, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.investment;

import jakarta.persistence.EntityNotFoundException;
import java.math.BigDecimal;
import java.text.MessageFormat;
import java.util.List;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.fxml.FXML;
import javafx.geometry.Pos;
import javafx.scene.control.TableCell;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.control.TextField;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.investment.Bond;
import org.moinex.service.BondService;
import org.moinex.service.I18nService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

@Controller
@NoArgsConstructor
public class ArchivedBondsController {
    @FXML private TableView<Bond> bondTableView;

    @FXML private TextField searchField;

    private List<Bond> archivedBonds;

    private BondService bondService;
    private I18nService i18nService;

    @Autowired
    public ArchivedBondsController(BondService bondService, I18nService i18nService) {
        this.bondService = bondService;
        this.i18nService = i18nService;
    }

    @FXML
    public void initialize() {
        loadArchivedBondsFromDatabase();

        configureTableView();

        updateBondTableView();

        // Add listener to the search field
        searchField
                .textProperty()
                .addListener((observable, oldValue, newValue) -> updateBondTableView());
    }

    @FXML
    private void handleUnarchive() {
        Bond selectedBond = bondTableView.getSelectionModel().getSelectedItem();

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_NO_BOND_SELECTED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys.BOND_DIALOG_NO_BOND_SELECTED_UNARCHIVE));
            return;
        }

        if (WindowUtils.showConfirmationDialog(
                MessageFormat.format(
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_DIALOG_CONFIRM_UNARCHIVE_TITLE),
                        selectedBond.getName()),
                i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_CONFIRM_UNARCHIVE_MESSAGE),
                i18nService.getBundle())) {
            try {
                bondService.unarchiveBond(selectedBond.getId());

                WindowUtils.showSuccessDialog(
                        i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_BOND_UNARCHIVED_TITLE),
                        MessageFormat.format(
                                i18nService.tr(
                                        Constants.TranslationKeys
                                                .BOND_DIALOG_BOND_UNARCHIVED_MESSAGE),
                                selectedBond.getName()));

                // Remove this bond from the list and update the table view
                archivedBonds.remove(selectedBond);
                updateBondTableView();
            } catch (EntityNotFoundException e) {
                WindowUtils.showErrorDialog(
                        i18nService.tr(
                                Constants.TranslationKeys.BOND_DIALOG_ERROR_UNARCHIVING_TITLE),
                        e.getMessage());
            }
        }
    }

    @FXML
    private void handleDelete() {
        Bond selectedBond = bondTableView.getSelectionModel().getSelectedItem();

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_NO_BOND_SELECTED_TITLE),
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_NO_BOND_SELECTED_DELETE));
            return;
        }

        // Prevent the removal of a bond with associated operations
        if (bondService.getOperationCountByBond(selectedBond.getId()) > 0) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_HAS_OPERATIONS_TITLE),
                    i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_HAS_OPERATIONS_MESSAGE));
            return;
        }

        if (WindowUtils.showConfirmationDialog(
                MessageFormat.format(
                        i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_CONFIRM_DELETE_TITLE),
                        selectedBond.getName()),
                i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_CONFIRM_DELETE_MESSAGE),
                i18nService.getBundle())) {
            try {
                bondService.deleteBond(selectedBond.getId());

                WindowUtils.showSuccessDialog(
                        i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_BOND_DELETED_TITLE),
                        MessageFormat.format(
                                i18nService.tr(
                                        Constants.TranslationKeys.BOND_DIALOG_BOND_DELETED_MESSAGE),
                                selectedBond.getName()));

                // Remove this bond from the list and update the table view
                archivedBonds.remove(selectedBond);
                updateBondTableView();
            } catch (EntityNotFoundException | IllegalStateException e) {
                WindowUtils.showErrorDialog(
                        i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_ERROR_DELETING_TITLE),
                        e.getMessage());
            }
        }
    }

    @FXML
    private void handleCancel() {
        Stage stage = (Stage) searchField.getScene().getWindow();
        stage.close();
    }

    private void loadArchivedBondsFromDatabase() {
        archivedBonds = bondService.getAllArchivedBonds();
    }

    private void updateBondTableView() {
        String searchText = searchField.getText().toLowerCase();

        bondTableView.getItems().clear();

        // Populate the table view
        if (searchText.isEmpty()) {
            bondTableView.getItems().setAll(archivedBonds);
        } else {
            archivedBonds.stream()
                    .filter(
                            b -> {
                                String name = b.getName().toLowerCase();
                                String symbol =
                                        b.getSymbol() != null ? b.getSymbol().toLowerCase() : "";
                                String type = b.getType().toString().toLowerCase();
                                String issuer =
                                        b.getIssuer() != null ? b.getIssuer().toLowerCase() : "";

                                return name.contains(searchText)
                                        || symbol.contains(searchText)
                                        || type.contains(searchText)
                                        || issuer.contains(searchText);
                            })
                    .forEach(bondTableView.getItems()::add);
        }

        bondTableView.refresh();
    }

    private void configureTableView() {
        TableColumn<Bond, Integer> idColumn = getBondIdTableColumn();

        TableColumn<Bond, String> nameColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_NAME));
        nameColumn.setCellValueFactory(
                param -> new SimpleStringProperty(param.getValue().getName()));

        TableColumn<Bond, String> symbolColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_SYMBOL));
        symbolColumn.setCellValueFactory(
                param -> new SimpleStringProperty(param.getValue().getSymbol()));

        TableColumn<Bond, String> typeColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_TYPE));
        typeColumn.setCellValueFactory(
                param ->
                        new SimpleStringProperty(
                                UIUtils.translateBondType(
                                        param.getValue().getType(), i18nService)));

        TableColumn<Bond, String> issuerColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_ISSUER));
        issuerColumn.setCellValueFactory(
                param -> new SimpleStringProperty(param.getValue().getIssuer()));

        TableColumn<Bond, String> quantityColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_QUANTITY));
        quantityColumn.setCellValueFactory(
                param -> {
                    Bond bond = param.getValue();
                    BigDecimal quantity = bondService.getCurrentQuantity(bond);
                    return new SimpleStringProperty(quantity.toString());
                });

        TableColumn<Bond, String> avgPriceColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_UNIT_PRICE));
        avgPriceColumn.setCellValueFactory(
                param -> {
                    Bond bond = param.getValue();
                    BigDecimal avgPrice = bondService.getAverageUnitPrice(bond);
                    return new SimpleStringProperty(UIUtils.formatCurrency(avgPrice));
                });

        TableColumn<Bond, String> investedValueColumn =
                new TableColumn<>(
                        i18nService.tr(Constants.TranslationKeys.BOND_TABLE_INVESTED_VALUE));
        investedValueColumn.setCellValueFactory(
                param -> {
                    Bond bond = param.getValue();
                    BigDecimal invested = bondService.getInvestedValue(bond);
                    return new SimpleStringProperty(UIUtils.formatCurrency(invested));
                });

        TableColumn<Bond, String> maturityDateColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_MATURITY_DATE));
        maturityDateColumn.setCellValueFactory(
                param -> {
                    Bond bond = param.getValue();
                    if (bond.getMaturityDate() != null) {
                        return new SimpleStringProperty(
                                UIUtils.formatDateForDisplay(bond.getMaturityDate(), i18nService));
                    }
                    return new SimpleStringProperty("-");
                });

        // Add the columns to the table view
        bondTableView.getColumns().add(idColumn);
        bondTableView.getColumns().add(nameColumn);
        bondTableView.getColumns().add(symbolColumn);
        bondTableView.getColumns().add(typeColumn);
        bondTableView.getColumns().add(issuerColumn);
        bondTableView.getColumns().add(quantityColumn);
        bondTableView.getColumns().add(avgPriceColumn);
        bondTableView.getColumns().add(investedValueColumn);
        bondTableView.getColumns().add(maturityDateColumn);
    }

    private TableColumn<Bond, Integer> getBondIdTableColumn() {
        TableColumn<Bond, Integer> idColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_ID));
        idColumn.setCellValueFactory(param -> new SimpleObjectProperty<>(param.getValue().getId()));

        // Align the ID column to the center
        idColumn.setCellFactory(
                column ->
                        new TableCell<>() {
                            @Override
                            protected void updateItem(Integer item, boolean empty) {
                                super.updateItem(item, empty);
                                if (item == null || empty) {
                                    setText(null);
                                } else {
                                    setText(item.toString());
                                    setAlignment(Pos.CENTER);
                                    setStyle("-fx-padding: 0;");
                                }
                            }
                        });
        return idColumn;
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/investment/view/ArchivedBondsController.kt
/*
 * Filename: ArchivedBondsController.kt (original filename: ArchivedBondsController.java)
 * Created on: January  3, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.investment.view

import jakarta.persistence.EntityNotFoundException
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.fxml.FXML
import javafx.geometry.Pos
import javafx.scene.control.TableColumn
import javafx.scene.control.TableView
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constants.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.investment.Bond
import org.moinex.service.PreferencesService
import org.moinex.service.investment.BondService
import org.springframework.stereotype.Controller
import java.text.MessageFormat

@Controller
class ArchivedBondsController(
    private val bondService: BondService,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var bondTableView: TableView<Bond>

    @FXML
    private lateinit var searchField: TextField

    private var archivedBonds: MutableList<Bond> = mutableListOf()

    @FXML
    fun initialize() {
        loadArchivedBondsFromDatabase()
        configureTableView()
        updateBondTableView()

        searchField.textProperty().addListener { _, _, _ ->
            updateBondTableView()
        }
    }

    @FXML
    private fun handleUnarchive() {
        val selectedBond = bondTableView.selectionModel.selectedItem

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_BOND_SELECTED_TITLE),
                preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_BOND_SELECTED_UNARCHIVE),
            )
            return
        }

        if (WindowUtils.showConfirmationDialog(
                MessageFormat.format(
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_CONFIRM_UNARCHIVE_TITLE),
                    selectedBond.name,
                ),
                preferencesService.translate(TranslationKeys.BOND_DIALOG_CONFIRM_UNARCHIVE_MESSAGE),
                preferencesService.bundle,
            )
        ) {
            runCatching {
                bondService.unarchiveBond(selectedBond.id!!)

                WindowUtils.showSuccessDialog(
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_BOND_UNARCHIVED_TITLE),
                    MessageFormat.format(
                        preferencesService.translate(TranslationKeys.BOND_DIALOG_BOND_UNARCHIVED_MESSAGE),
                        selectedBond.name,
                    ),
                )

                archivedBonds.remove(selectedBond)
                updateBondTableView()
            }.onFailure { e ->
                if (e is EntityNotFoundException) {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.BOND_DIALOG_ERROR_UNARCHIVING_TITLE),
                        e.message ?: "",
                    )
                } else {
                    throw e
                }
            }
        }
    }

    @FXML
    private fun handleDelete() {
        val selectedBond = bondTableView.selectionModel.selectedItem

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_BOND_SELECTED_TITLE),
                preferencesService.translate(TranslationKeys.BOND_DIALOG_NO_BOND_SELECTED_DELETE),
            )
            return
        }

        if (bondService.getOperationCountByBond(selectedBond.id!!) > 0) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.BOND_DIALOG_HAS_OPERATIONS_TITLE),
                preferencesService.translate(TranslationKeys.BOND_DIALOG_HAS_OPERATIONS_MESSAGE),
            )
            return
        }

        if (WindowUtils.showConfirmationDialog(
                MessageFormat.format(
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_CONFIRM_DELETE_TITLE),
                    selectedBond.name,
                ),
                preferencesService.translate(TranslationKeys.BOND_DIALOG_CONFIRM_DELETE_MESSAGE),
                preferencesService.bundle,
            )
        ) {
            runCatching {
                bondService.deleteBond(selectedBond.id!!)

                WindowUtils.showSuccessDialog(
                    preferencesService.translate(TranslationKeys.BOND_DIALOG_BOND_DELETED_TITLE),
                    MessageFormat.format(
                        preferencesService.translate(TranslationKeys.BOND_DIALOG_BOND_DELETED_MESSAGE),
                        selectedBond.name,
                    ),
                )

                archivedBonds.remove(selectedBond)
                updateBondTableView()
            }.onFailure { e ->
                if (e is EntityNotFoundException || e is IllegalStateException) {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.BOND_DIALOG_ERROR_DELETING_TITLE),
                        e.message ?: "",
                    )
                } else {
                    throw e
                }
            }
        }
    }

    @FXML
    private fun handleCancel() {
        (searchField.scene.window as Stage).close()
    }

    private fun loadArchivedBondsFromDatabase() {
        archivedBonds = bondService.getAllArchivedBonds().toMutableList()
    }

    private fun updateBondTableView() {
        val searchText = searchField.text.lowercase()

        bondTableView.items.clear()

        if (searchText.isEmpty()) {
            bondTableView.items.setAll(archivedBonds)
        } else {
            archivedBonds
                .filter { b ->
                    val name = b.name.lowercase()
                    val symbol = b.symbol?.lowercase() ?: ""
                    val type = b.type.toString().lowercase()
                    val issuer = b.issuer?.lowercase() ?: ""

                    name.contains(searchText) ||
                        symbol.contains(searchText) ||
                        type.contains(searchText) ||
                        issuer.contains(searchText)
                }.forEach { bondTableView.items.add(it) }
        }

        bondTableView.refresh()
    }

    private fun configureTableView() {
        val idColumn =
            TableColumn<Bond, Int>(
                preferencesService.translate(TranslationKeys.BOND_TABLE_ID),
            ).apply {
                setCellValueFactory { param -> SimpleObjectProperty(param.value.id) }
                UIUtils.alignTableColumn(this, Pos.CENTER)
            }

        val nameColumn = TableColumn<Bond, String>(preferencesService.translate(TranslationKeys.BOND_TABLE_NAME))
        nameColumn.setCellValueFactory { param -> SimpleStringProperty(param.value.name) }

        val symbolColumn = TableColumn<Bond, String>(preferencesService.translate(TranslationKeys.BOND_TABLE_SYMBOL))
        symbolColumn.setCellValueFactory { param -> SimpleStringProperty(param.value.symbol) }

        val typeColumn = TableColumn<Bond, String>(preferencesService.translate(TranslationKeys.BOND_TABLE_TYPE))
        typeColumn.setCellValueFactory { param ->
            SimpleStringProperty(UIUtils.translateBondType(param.value.type))
        }

        val issuerColumn = TableColumn<Bond, String>(preferencesService.translate(TranslationKeys.BOND_TABLE_ISSUER))
        issuerColumn.setCellValueFactory { param -> SimpleStringProperty(param.value.issuer) }

        val quantityColumn =
            TableColumn<Bond, String>(preferencesService.translate(TranslationKeys.BOND_TABLE_QUANTITY))
        quantityColumn.setCellValueFactory { param ->
            val bond = param.value
            val quantity = bondService.getCurrentQuantity(bond)
            SimpleStringProperty(quantity.toString())
        }

        val avgPriceColumn =
            TableColumn<Bond, String>(preferencesService.translate(TranslationKeys.BOND_TABLE_UNIT_PRICE))
        avgPriceColumn.setCellValueFactory { param ->
            val bond = param.value
            val avgPrice = bondService.getAverageUnitPrice(bond)
            SimpleStringProperty(UIUtils.formatCurrency(avgPrice))
        }

        val investedValueColumn =
            TableColumn<Bond, String>(preferencesService.translate(TranslationKeys.BOND_TABLE_INVESTED_VALUE))
        investedValueColumn.setCellValueFactory { param ->
            val bond = param.value
            val invested = bondService.getTotalInvestedValue(bond)
            SimpleStringProperty(UIUtils.formatCurrency(invested))
        }

        val maturityDateColumn =
            TableColumn<Bond, String>(preferencesService.translate(TranslationKeys.BOND_MATURITY_DATE))
        maturityDateColumn.setCellValueFactory { param ->
            val bond = param.value
            if (bond.maturityDate != null) {
                SimpleStringProperty(UIUtils.formatDateForDisplay(bond.maturityDate))
            } else {
                SimpleStringProperty("-")
            }
        }

        bondTableView.columns.add(idColumn)
        bondTableView.columns.add(nameColumn)
        bondTableView.columns.add(symbolColumn)
        bondTableView.columns.add(typeColumn)
        bondTableView.columns.add(issuerColumn)
        bondTableView.columns.add(quantityColumn)
        bondTableView.columns.add(avgPriceColumn)
        bondTableView.columns.add(investedValueColumn)
        bondTableView.columns.add(maturityDateColumn)
    }
}
